yogstation/code/modules/liquids/liquid_groups.dm

```diff
@@ -0,0 +1,1024 @@
+		source_turf.liquids.liquid_group.transfer_reagents_to_secondary_group(source_turf.liquids, Z_turf_below.liquids)
+
+		var/obj/splashy = new /obj/effect/temp_visual/liquid_splash(Z_turf_below)
+		if(!QDELETED(Z_turf_below.liquids?.liquid_group))
+			splashy.color = Z_turf_below.liquids.liquid_group.group_color
+		return FALSE
+
+	if(QDELETED(new_turf.liquids) && !istype(new_turf, /turf/open/openspace) && !isspaceturf(new_turf) && !istype(new_turf, /turf/open/floor/plating/ocean) && source_turf.turf_height == new_turf.turf_height) // no space turfs, or oceans turfs, also don't attempt to spread onto a turf that already has liquids wastes processing time
+		if(reagents_per_turf < LIQUID_HEIGHT_DIVISOR)
+			return FALSE
+		if(!length(members))
+			return FALSE
+		reagents_per_turf = total_reagent_volume / length(members)
+		expected_turf_height = CEILING(reagents_per_turf, 1) / LIQUID_HEIGHT_DIVISOR
+		new_turf.liquids = new(new_turf, src)
+		new_turf.liquids.alpha = group_alpha
+		check_edges(new_turf)
+
+		var/obj/splashy = new /obj/effect/temp_visual/liquid_splash(new_turf)
+		if(!QDELETED(new_turf.liquids?.liquid_group))
+			splashy.color = new_turf.liquids.liquid_group.group_color
+
+		water_rush(new_turf, source_turf)
+
+	else if(!QDELETED(source_turf?.liquids?.liquid_group) && !QDELETED(new_turf?.liquids?.liquid_group) && new_turf.liquids.liquid_group != source_turf.liquids.liquid_group && source_turf.turf_height == new_turf.turf_height && new_turf.liquids.liquid_group.can_merge)
+		merge_group(new_turf.liquids.liquid_group)
+		return FALSE
+	else if(source_turf.turf_height != new_turf.turf_height)
+		if(new_turf.turf_height < source_turf.turf_height) // your going down
+			if(QDELETED(new_turf.liquids))
+				new_turf.liquids = new(new_turf)
+			if(new_turf.turf_height + new_turf.liquids.liquid_group.expected_turf_height < source_turf.turf_height)
+				var/obj/effect/abstract/liquid_turf/turf_liquids = new_turf.liquids
+				trans_to_seperate_group(turf_liquids.liquid_group.reagents, reagents_per_turf, source_turf)
+				turf_liquids.liquid_group.process_group()
+		else if(source_turf.turf_height < new_turf.turf_height)
+			if(source_turf.turf_height + expected_turf_height < new_turf.turf_height)
+				return
+			if(QDELETED(new_turf.liquids))
+				new_turf.liquids = new(new_turf)
+			var/obj/effect/abstract/liquid_turf/turf_liquids = new_turf.liquids
+			trans_to_seperate_group(turf_liquids.liquid_group.reagents, 10, source_turf) //overflows out
+			turf_liquids.liquid_group.process_group()
+
+	return TRUE
+
+/datum/liquid_group/proc/water_rush(turf/new_turf, turf/source_turf)
+	var/direction = get_dir(source_turf, new_turf)
+	for(var/atom/movable/target_atom in new_turf)
+		if(!target_atom.anchored && !target_atom.pulledby && !isobserver(target_atom) && (target_atom.move_resist < INFINITY))
+			reagents.reaction(target_atom, TOUCH, (reagents_per_turf * 0.5))
+			if(expected_turf_height < LIQUID_ANKLES_LEVEL_HEIGHT)
+				return
```

Comment:
probably worth putting reagents.reaction below the check
if the liquid isn't enough to push them, chances are it isn't gonna react with them

also, the react with reagents should probably still happen, even if the target is anchored or being pulled, or has infinity move resist

also also, magboots should probably resist this, checking heavy gravity is what it does